main.go

```diff
@@ -26,3 +47,209 @@ func main() {
+	// errors for debugging and understanding.
+	return ss.Close()
+}
+
+// AfterApply on CLI runs before AfterApply of any commands, creating a
+// GitCredential from stdin and opening a connection to DBus and creating
+// a session with the secret service. These two resources are bound to
+// the kong context to make them available to the command Run methods.
+func (cmd *CLI) AfterApply(kctx *kong.Context) error {
+	// Create a GitCredential from the lines on stdin. See
+	// git-credential(1) for the format.
+	// https://git-scm.com/docs/git-credential#IOFMT
+	gc := &GitCredential{}
+	if err := gc.Unmarshal(os.Stdin); err != nil {
+		return err
+	}
+	kctx.Bind(gc)
+
+	// Open a DBus connection and create a session with the secret service.
+	// https://specifications.freedesktop.org/secret-service/latest/
+	ss, err := NewSecretService()
+	if err != nil {
+		return err
+	}
+	kctx.Bind(ss)
+
+	return nil
+}
+
+// AfterApply validates the input credential fields for a get command.
+func (cmd *CmdGet) AfterApply(gc *GitCredential) error {
+	cmd.validate(gc.Protocol != "", "protocol")
+	cmd.validate(gc.Host != "" || gc.Path != "", "host or path")
+	return cmd.err
+}
+
+// Run executes the credential helper "get" operation.
+//
+// The "get" operation is specified by the [gitcredentials] documentation, amd
+// exists to look up a password and/or other secret material to access a remote
+// git repository, previously stored with a "store" operation.
+//
+// [gitcredentials]: https://git-scm.com/docs/gitcredentials
+func (cmd *CmdGet) Run(gc *GitCredential, ss *SecretService) error {
+	secret, err := ss.Get(makeAttrs(gc))
+	if err != nil {
+		return err
+	}
+
+	if secret == "" {
+		return nil
+	}
+
+	if err := parseSecretVal(secret, gc); err != nil {
+		return err
+	}
+
+	return gc.Marshal(os.Stdout)
+}
+
+// AfterApply validates the input credential fields for a store command.
+func (cmd *CmdStore) AfterApply(gc *GitCredential) error {
+	cmd.validate(gc.Protocol != "", "protocol")
+	cmd.validate(gc.Username != "", "username")
+	cmd.validate(gc.Password != "", "password")
+	cmd.validate(gc.Host != "" || gc.Path != "", "host or path")
+	return cmd.err
+}
+
+// Run executes the credential helper "store" operation.
+//
+// The "store" operation is specified by the [gitcredentials] documentation,
+// amd exists to store a password and/or other secret material needed to access
+// a remote git repository.
+//
+// [gitcredentials]: https://git-scm.com/docs/gitcredentials
+func (cmd *CmdStore) Run(gc *GitCredential, ss *SecretService) error {
+	return ss.Store(makeLabel(gc), makeAttrs(gc), makeSecretVal(gc))
+}
+
+// AfterApply validates the input credential fields for a erase command.
+func (cmd *CmdErase) AfterApply(gc *GitCredential) error {
+	cmd.validate(gc.Protocol != "", "protocol")
+	cmd.validate(gc.Username != "", "username")
+	cmd.validate(gc.Host != "", "host")
+	cmd.validate(gc.Path != "", "path")
```

Comment:
interesting that how and path are required here where in
all other places its host or path

Reply:
Yeah. I've changed that. This is what git-credential-libsecret is doing, but I now think it's wrong and I won't copy that: https://github.com/git/git/blob/159f2d50e75c17382c9f4eb7cbda671a6fa612d1/contrib/credential/libsecret/git-credential-libsecret.c#L255